src/backend/vulkan/src/xr.rs

```diff
@@ -0,0 +1,330 @@
+use ash::vk::{self, Handle};
+use hal::image;
+use std::{
+    ffi::{CStr, CString},
+    sync::{atomic::Ordering, Mutex},
+};
+
+#[cfg(feature = "use-openxr")]
+use once_cell::sync::Lazy;
+use std::sync::atomic::AtomicBool;
+
+use crate::{conv, native};
+
+#[cfg(feature = "use-openxr")]
+pub(crate) struct Instance {
+    instance: openxr::Instance,
+    system: openxr::SystemId,
+    required_device_extension_properties: Vec<CString>,
+    required_instance_extension_properties: Vec<CString>,
+    requirements: openxr::vulkan::Requirements,
+
+    session: Option<openxr::Session<openxr::Vulkan>>,
+    frame_waiter: Option<openxr::FrameWaiter>,
+    frame_stream: Option<openxr::FrameStream<openxr::Vulkan>>,
+    space: Option<openxr::Space>,
+    // FIXME add field for all -validations state, that'd track separate parts
+}
+
+#[cfg(feature = "use-openxr")]
+impl std::fmt::Debug for Instance {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "OpenXR[...]")
+    }
+}
+
+#[cfg(feature = "use-openxr")]
+pub(crate) static INSTANCE: Lazy<Mutex<Option<Instance>>> = Lazy::new(|| Mutex::new(None));
```

Comment:
We shouldn't need this. Instead, one of the two approaches should be taken:
  1. this data is a part of the main `Instance`
  2. it's managed by the user, and gfx-backend-vulkan just understands it and knows how to cooperate

Reply:
This was the biggest hurdle I got into. Been a while since I made this decision, but:
* avoid changing the `hal` structs (which would require changes to all dependents)
* could not access instance from device-specific code (`impl hal::Backend for Backend`), but now it sems that `Arc<RawDevice>` contains `Arc<RawInstance>`, so in theory the required openxr data could be stored in those?

Reply:
yes, we should store the openxr data in `RawDevice` or `RawInstance` if it's needed in multiple places.